optimum/habana/transformers/models/gpt_neox/modeling_gpt_neox.py

```diff
@@ -457,19 +457,11 @@ def prepare_inputs_for_generation(
         return model_inputs
 
 
-def gaudi_gpt_neox_rotary_embedding_set_cos_sin_cache(self, seq_len, device, dtype):
-    self.max_seq_len_cached = seq_len
-    t = torch.arange(self.max_seq_len_cached, device=device, dtype=self.inv_freq.dtype)
-
-    freqs = torch.outer(t, self.inv_freq)
-    # Different from paper, but it uses a different permutation in order to obtain the same calculation
-    emb = torch.cat((freqs, freqs), dim=-1)
-    self.cos_cached = emb.cos()
-    self.sin_cached = emb.sin()
-
-
 def apply_customized_rope(q, k, cos, sin, position_ids, training=True):
     if q.device.type == "hpu" and FusedRoPE:
-        return apply_customized_rope_module(q, k, cos, sin, position_ids, training)
+        if training:
+            return apply_customized_rope_module(q.to(torch.float), k.to(torch.float), cos, sin, position_ids, training)
```

Comment:
Thanks @yafshar for noticing this. 
Can this be added here :https://github.com/huggingface/optimum-habana/blob/04901ae0746548856463e733a04eb89f40812903/optimum/habana/transformers/models/modeling_all_models.py#L81  (L83-84) to make the training segment default to torch.float32 ? Then model wise changes would not be required. It is strange that there are some instabilities as by default it should be torch .float32 , and there is no forced typecasts in the commits before my PR.  However I realise that forced typecasting may be needed for other models as well in training phase.

Reply:
@abhilash1910, yes, technically, it should be added as you suggested. However, I do not know if this instability affects other models. It affects performance. So, before applying it to all models, we should do extra checks.

Reply:
sounds good, thanks